File: 3548_Equal_Sum_Grid_Partition_II/solution.py
class Solution:
    def canPartitionGrid(self, grid: List[List[int]]) -> bool:
        def solve(g):
            n, m = len(g), len(g[0])
            bot, top = 0, 0
            bf = [0] * 100001
            tf = [0] * 100001
            for i in range(n):
                for j in range(m):
                    bot += g[i][j]
                    bf[g[i][j]] += 1
            for i in range(n - 1):
                for j in range(m):
                    bot -= g[i][j]
                    bf[g[i][j]] -= 1
                    top += g[i][j]
                    tf[g[i][j]] += 1
                if top == bot: return True
                dt = top - bot
                if 0 < dt <= 100000:
                    h, w = i + 1, m
                    if h > 1 and w > 1:
                        if tf[dt]: return True
                    elif h > 1 and w == 1:
                        if g[0][0] == dt or g[i][0] == dt: return True
                    elif h == 1 and w > 1:
                        if g[0][0] == dt or g[0][w - 1] == dt: return True
                db = bot - top
                if 0 < db <= 100000:
                    h, w = n - i - 1, m
                    if h > 1 and w > 1:
                        if bf[db]: return True
                    elif h > 1 and w == 1:
                        if g[i + 1][0] == db or g[n - 1][0] == db: return True
                    elif h == 1 and w > 1:
                        if g[n - 1][0] == db or g[n - 1][w - 1] == db: return True
            return False

        if solve(grid): return True
        n, m = len(grid), len(grid[0])
        mat = [[grid[i][j] for i in range(n)] for j in range(m)]
        return solve(mat)

File: 3548_Equal_Sum_Grid_Partition_II/solution.c
static bool solve(int** g, int n, int m) {
    long long bot = 0, top = 0;
    int* bf = calloc(100001, sizeof(int)); int* tf = calloc(100001, sizeof(int));

    for (int i = 0; i < n; i++) for (int j = 0; j < m; j++) 
    { bot += g[i][j]; bf[g[i][j]]++; }

    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < m; j++) {
            bot -= g[i][j];
            bf[g[i][j]]--;
            top += g[i][j];
            tf[g[i][j]]++;
        }
        if (top == bot) { free(bf); free(tf); return true; }

        long long dt = top - bot;
        if (dt > 0 && dt <= 100000) {
            int h = i + 1, w = m;
            if (h > 1 && w > 1) {
                if (tf[dt]) { free(bf); free(tf); return true; }
            } else if (h > 1 && w == 1) {
                if (g[0][0] == dt || g[i][0] == dt) { free(bf); free(tf); return true; }
            } else if (h == 1 && w > 1) {
                if (g[0][0] == dt || g[0][w - 1] == dt) { free(bf); free(tf); return true; }
            }
        }

        long long db = bot - top;
        if (db > 0 && db <= 100000) {
            int h = n - i - 1, w = m;
            if (h > 1 && w > 1) {
                if (bf[db]) { free(bf); free(tf); return true; }
            } else if (h > 1 && w == 1) {
                if (g[i + 1][0] == db || g[n - 1][0] == db) { free(bf); free(tf); return true; }
            } else if (h == 1 && w > 1) {
                if (g[n - 1][0] == db || g[n - 1][w - 1] == db) { free(bf); free(tf); return true; }
            }
        }
    }
    free(bf);
    free(tf);
    return false;
}

bool canPartitionGrid(int** grid, int gridSize, int* gridColSize) {
    int n = gridSize, m = gridColSize[0];

    if (solve(grid, n, m)) return true;

    int** mat = malloc(m * sizeof(int*));
    for (int j = 0; j < m; j++) {
        mat[j] = malloc(n * sizeof(int));
        for (int i = 0; i < n; i++) mat[j][i] = grid[i][j];
    }

    bool res = solve(mat, m, n);
    for (int j = 0; j < m; j++) free(mat[j]);
    free(mat);
    return res;
}